File: benchmarks/pandas/bench_value_counts_opts.py
"""Benchmark: value_counts with options — normalize=True, ascending=True, dropna=False."""
import json
import time
import pandas as pd
import numpy as np

ROWS = 100_000
WARMUP = 3
ITERATIONS = 20

data = [None if i % 500 == 0 else f"cat_{i % 50}" for i in range(ROWS)]
s = pd.Series(data)

for _ in range(WARMUP):
    s.value_counts(normalize=True)
    s.value_counts(ascending=True)
    s.value_counts(dropna=False)
    s.value_counts(normalize=True, ascending=True, dropna=False)

start = time.perf_counter()
for _ in range(ITERATIONS):
    s.value_counts(normalize=True)
    s.value_counts(ascending=True)
    s.value_counts(dropna=False)
    s.value_counts(normalize=True, ascending=True, dropna=False)
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "value_counts_opts",
    "mean_ms": total / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total,
}))

File: benchmarks/tsb/bench_value_counts_opts.ts
/**
 * Benchmark: valueCounts with options — normalize=true, ascending=true, dropna=false.
 * Outputs JSON: {"function": "value_counts_opts", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Series, valueCounts } from "../../src/index.ts";

const ROWS = 100_000;
const WARMUP = 3;
const ITERATIONS = 20;

const data = Array.from({ length: ROWS }, (_, i) => {
  if (i % 500 === 0) return null;
  return `cat_${i % 50}`;
});
const s = new Series({ data });

for (let i = 0; i < WARMUP; i++) {
  valueCounts(s, { normalize: true });
  valueCounts(s, { ascending: true });
  valueCounts(s, { dropna: false });
  valueCounts(s, { normalize: true, ascending: true, dropna: false });
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  valueCounts(s, { normalize: true });
  valueCounts(s, { ascending: true });
  valueCounts(s, { dropna: false });
  valueCounts(s, { normalize: true, ascending: true, dropna: false });
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "value_counts_opts",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);
